packages/dataviews/src/view-actions.tsx

```diff
@@ -58,10 +98,15 @@ function ViewTypeMenu( { view, onChangeView, supportedLayouts } ) {
 						name="view-actions-available-view"
 						checked={ availableView.type === view.type }
 						hideOnClick
-						onChange={ ( e ) => {
+						onChange={ ( e: ChangeEvent< HTMLInputElement > ) => {
```

Comment:
The main challenge of this PR is this function: Basically how to keep the "view" object valid (correct type) when switching view types.

Reply:
Transitional states are difficult, this seems to highlight complexity that wasn't apparent, although it seems like the complexity isn't related to types as much as it is to application states. What should that layout look like?

If the layout properties are truly optional then this makes sense and may be better than the previous behavior where dataviews would maintain layout properties when switching views.

---


I know this isn't your question, but I try to avoid type assertions like these and would typically do something like this here:

```js
onChange={ ( e: ChangeEvent< HTMLInputElement > ) => {
	switch ( e.target.value ) {
		case 'list':
		case 'grid':
		case 'table':
			return onChangeView( {
				...view,
				type: e.target.value,
				layout: {},
			} );
	}
	throw new Error( 'Invalid dataview' );
} }
```

Reply:
I agree the complexity is on the app level: "how to switch view types when the view types can be different" and we don't have a good answer yet.